### PR TITLE
fix(x/ecocredit)!: add batch balance migration

### PR DESCRIPTION
**Pull request by aleem1314**
## Description

ref: #1660 

Script to get all affected account [link](https://colab.research.google.com/drive/1pZU8HnWFce94BCFWwSUxxaJeEpa3X_QX?usp=sharing).
Also did manual testing with mainnet export.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: app/upgrades/v5_0/upgrade_test.go [hunk 1]
```diff
@@ -30,6 +30,8 @@ func (suite *UpgradeTestSuite) TestV5Upgrade() {
 	suite.Setup()
 
 	suite.Ctx = suite.Ctx.WithBlockHeight(dummyUpgradeHeight - 1)
+	suite.Ctx = suite.Ctx.WithChainID("testing-1")
```
Comment (aleem1314) [inline app/upgrades/v5_0/upgrade_test.go:33]: Setting custom chain-id to skip manual state migrations.

File: x/ecocredit/base/keeper/msg_send.go [hunk 2]
```diff
@@ -49,125 +49,212 @@ func (k Keeper) sendEcocredits(sdkCtx sdk.Context, credit *types.MsgSend_SendCre
 	}
 	precision := creditType.Precision
 
-	batchSupply, err := k.stateStore.BatchSupplyTable().Get(ctx, batch.Key)
+	creditDecs, err := utils.GetNonNegativeFixedDecs(precision, credit.TradableAmount, credit.RetiredAmount)
```
Comment (aleem1314) [inline x/ecocredit/base/keeper/msg_send.go:52]: Refactored `sendEcocredits` to support self-transfer.

Comment (ryanchristo) [reply x/ecocredit/base/keeper/msg_send.go:52]: We should probably still error in basic validation if a user attempts to send credits to themselves but the refactor makes sense either way. We could allow self-transfer with these changes but I think better not to have the user spend gas for what would most likely be a user error.

Comment (ryanchristo) [reply x/ecocredit/base/keeper/msg_send.go:52]: I did a bit of an audit and additional refactoring based off your branch, which could be merged into your branch or considered for a followup. I think it improves upon what you fixed here and makes things a bit more readable and consistent and avoids passing storage as a parameter. https://github.com/regen-network/regen-ledger/pull/1680

File: x/ecocredit/migrations/v3/state_test.go [hunk 3]
```diff
@@ -100,3 +75,89 @@ func TestMigrations(t *testing.T) {
 
 	require.Equal(t, params.AllowedClassCreators, expected)
 }
+
+func TestBasketBalanceMigration(t *testing.T) {
```
Comment (ryanchristo) [inline x/ecocredit/migrations/v3/state_test.go:79]:
```suggestion
func TestBatchBalanceMigration(t *testing.T) {
```

File: x/ecocredit/migrations/v3/state_test.go [hunk 4]
```diff
@@ -100,3 +75,89 @@ func TestMigrations(t *testing.T) {
...
+
+	issuer := sdk.MustAccAddressFromBech32("regen1v2ncquer9r2ytlkxh2djmmsq3e8we6rjc9snfn")
+	creditHolder := sdk.MustAccAddressFromBech32("regen1l8v5nzznewg9cnfn0peg22mpysdr3a8jcm4p8v")
+
+	ormdb, err := ormdb.NewModuleDB(&ecocredit.ModuleSchema, ormdb.ModuleDBOptions{})
+	require.NoError(t, err)
+	baseStore, err := baseapi.NewStateStore(ormdb)
+	require.NoError(t, err)
+
+	err = baseStore.BatchTable().Save(sdkCtx, &baseapi.Batch{
```
Comment (ryanchristo) [inline x/ecocredit/migrations/v3/state_test.go:91]:
```suggestion
	err = baseStore.BatchTable().Insert(sdkCtx, &baseapi.Batch{
```

File: x/ecocredit/migrations/v3/state_test.go [hunk 5]
```diff
@@ -100,3 +75,89 @@ func TestMigrations(t *testing.T) {
...
+	err = baseStore.BatchBalanceTable().Save(sdkCtx, &baseapi.BatchBalance{
+		BatchKey:       1,
+		Address:        creditHolder,
+		TradableAmount: "0.00",
+		RetiredAmount:  "0",
+		EscrowedAmount: "0",
+	})
+	require.NoError(t, err)
+
+	err = baseStore.BatchBalanceTable().Save(sdkCtx, &baseapi.BatchBalance{
```
Comment (ryanchristo) [inline x/ecocredit/migrations/v3/state_test.go:107]:
```suggestion
	err = baseStore.BatchBalanceTable().Insert(sdkCtx, &baseapi.BatchBalance{
```

File: x/ecocredit/migrations/v3/state_test.go [hunk 6]
```diff
@@ -100,3 +75,89 @@ func TestMigrations(t *testing.T) {
...
+		BatchKey:       1,
+		Address:        issuer,
+		TradableAmount: "799.95",
+		RetiredAmount:  "100",
+		EscrowedAmount: "100",
+	})
+	require.NoError(t, err)
+
+	err = baseStore.BatchSupplyTable().Save(sdkCtx, &baseapi.BatchSupply{
```
Comment (ryanchristo) [inline x/ecocredit/migrations/v3/state_test.go:116]:
```suggestion
	err = baseStore.BatchSupplyTable().Insert(sdkCtx, &baseapi.BatchSupply{
```

**Review by ryanchristo — APPROVED** [review]
tACK

```
git checkout v4.1.5
```

```
~/Desktop/testnet_from_export.sh -e ~/Downloads/mainnet.json
```

```
./build/regen config chain-id regen-1
```

```
./build/regen config broadcast-mode block
```

```
./build/regen config keyring-backend test
```

```
./build/regen tx gov submit-proposal software-upgrade v5.0 --title test --description test --upgrade-height 8139285 --from test
```

```
./build/regen tx gov deposit 21 200000000uregen --from test
```

```
./build/regen tx gov vote 21 yes --from test
```

```
./build/regen q gov proposal 21
```

```
git checkout aleem/msg-send-migrations
```

```
make build
```

```
./build/regen start --fast_sync=false --minimum-gas-prices 0uregen
```

```
./build/regen export >> export.json
```

```
./build/regen validate-genesis export.json
```

**Pull request merged**